File: WEEK20/DFS_타겟 넘버.py
# DFS - 타겟 넘버

def solution(numbers, target):
    answer = 0
    
    def depth_first(idx, value):
        nonlocal answer
        
        if idx == len(numbers):
            if value == target:
                answer += 1
            return
        else:
            depth_first(idx + 1, value + numbers[idx])
            depth_first(idx + 1, value - numbers[idx])
    
    depth_first(0, 0)
    
    return answer

File: WEEK22/DFS_타겟 넘버.js
function solution(numbers, target) {
  var answer = 0;

  function depthFirst(index, value) {
    if (index === numbers.length) {
      if (value === target) {
        answer++;
      }

      return;
    }

    depthFirst(index + 1, value + numbers[index]);
    depthFirst(index + 1, value - numbers[index]);
  }

  depthFirst(0, 0);

  return answer;
}

console.log(solution([1, 1, 1, 1, 1], 3));
